libc/zipos/mmap.c

```diff
@@ -0,0 +1,79 @@
+│                                                                              │
+│ Permission to use, copy, modify, and/or distribute this software for         │
+│ any purpose with or without fee is hereby granted, provided that the         │
+│ above copyright notice and this permission notice appear in all copies.      │
+│                                                                              │
+│ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL                │
+│ WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED                │
+│ WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE             │
+│ AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL         │
+│ DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR        │
+│ PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER               │
+│ TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR             │
+│ PERFORMANCE OF THIS SOFTWARE.                                                │
+╚─────────────────────────────────────────────────────────────────────────────*/
+#include "libc/calls/calls.h"
+#include "libc/calls/struct/iovec.h"
+#include "libc/dce.h"
+#include "libc/errno.h"
+#include "libc/intrin/likely.h"
+#include "libc/intrin/strace.internal.h"
+#include "libc/sysv/consts/map.h"
+#include "libc/sysv/consts/prot.h"
+#include "libc/sysv/errfuns.h"
+#include "libc/zipos/zipos.internal.h"
+
+#define IP(X)  (intptr_t)(X)
+#define VIP(X) (void *)IP(X)
+
+/**
+ * Map zipos file into memory. See mmap.
+ *
+ * @param addr should be 0 or a compatible address
+ * @param size must be >0 and will be rounded up to FRAMESIZE
+ *     automatically.
+ * @param prot can have PROT_READ/PROT_WRITE/PROT_EXEC/PROT_NONE/etc.
+ * @param flags cannot have `MAP_SHARED` or `MAP_ANONYMOUS`, there is
+ *     no actual file backing for zipos files. `MAP_SHARED` could be
+ *     simulated for non-writable mappings, but that would require
+ *     tracking zipos mappings to prevent making it PROT_WRITE.
+ * @param h is a zip store object
+ * @param off specifies absolute byte index of h's file for mapping,
+ *     it does not need to be 64kb aligned.
+ * @return virtual base address of new mapping, or MAP_FAILED w/ errno
+ */
+void *__zipos_mmap(void *addr, size_t size, int prot, int flags,
+                   struct ZiposHandle *h, int64_t off) {
+  if (VERY_UNLIKELY(!!(flags & MAP_ANONYMOUS))) {
+    STRACE("MAP_ANONYMOUS zipos mismatch");
+    return VIP(einval());
+  }
+
+  // MAP_SHARED for non-writeable pages could be implemented, but would require
+  // keeping track of zipos pages to prevent mprotect(addr,len,PROT_WRITE)
+  if (VERY_UNLIKELY(!!(flags & MAP_SHARED))) {
+    STRACE("MAP_SHARED on zipos");
+    return VIP(eacces());
+  }
+
+  const int tempProt = !IsXnu() ? prot | PROT_WRITE : PROT_WRITE;
+  void *outAddr = mmap(addr, size, tempProt, flags | MAP_ANONYMOUS, -1, 0);
```

Comment:
Passing flags here is definitely the wrong thing to do. I would say be much stricter about flags. For example:

```c
if (!(flags == (MAP_FILE | MAP_PRIVATE) ||
      (flags == MAP_FILE && (~prot & PROT_WRITE)))) {
  STRACE("zipos mappings currently only support MAP_PRIVATE or non-writabl0e MAP_FILE");
  return einval();
}
```

Reply:
One additional note: You would want to support MAP_FIXED and MAP_FIXED_NOREPLACE. Those are the only other flags I could foresee being used. All other flags should be disallowed.